src/scaffold/admission_request.rs

```diff
@@ -0,0 +1,801 @@
+        if !catalog_dir.exists() {
+            std::fs::create_dir_all(catalog_dir)
+                .map_err(|err| anyhow!("failed to create cache directory: {err:?}"))?;
+        }
+
+        let file = File::create(catalog_path)
+            .map_err(|err| anyhow!("failed to create resource catalog file: {err:?}"))?;
+        serde_json::to_writer(&file, self)
+            .map_err(|err| anyhow!("failed to write resource catalog: {err:?}"))?;
+
+        Ok(())
+    }
+
+    pub fn lookup(
+        &self,
+        gvk: &kube::api::GroupVersionKind,
+    ) -> Option<&k8s_openapi::apimachinery::pkg::apis::meta::v1::APIResource> {
+        self.resources.get(&Self::gvk_to_string(gvk))
+    }
+
+    /// Refresh the catalog by querying the Kubernetes API server.
+    /// This applies only if the catalog was built from the cache.
+    pub async fn refresh<F, Fut>(&mut self, build_kubeclient: F) -> Result<()>
+    where
+        F: FnOnce() -> Fut + Clone,
+        Fut: Future<Output = Result<kube::Client>>,
+    {
+        match self.restored_from {
+            ApiResourceCatalogRestoredFrom::Cache => {
+                let client = build_kubeclient().await?;
+                let fresh_catalog = Self::build(client).await?;
+
+                self.resources.clear();
+                self.resources.extend(fresh_catalog.resources);
+                self.restored_from = ApiResourceCatalogRestoredFrom::ApiServer;
+            }
+            _ => {
+                // Nothing to do
+            }
+        }
+        Ok(())
+    }
+
+    /// Convert a GroupVersionKind to a string. This is used as a key in the catalog.
+    /// This proves to be faster to implement than doing a custom serialization/deserialization
+    /// for the GroupVersionKind struct.
+    ///
+    /// All the fields of GroupVersionKind are joined by a pipe character, which is a symbol
+    /// not accepted by Kubernetes.
+    fn gvk_to_string(gvk: &kube::api::GroupVersionKind) -> String {
+        format!("{}|{}|{}", gvk.group, gvk.version, gvk.kind)
+    }
+}
+
+// Build a client to interact with the Kubernetes API server.
+// The scaffold command must be snappy, we don't want it to get stuck
+// waiting for the connection to Kubernetes to be established.
+// Because of that we set a connection timeout of 1 second.
+async fn kube_client() -> Result<kube::Client> {
+    let mut config = kube::Config::infer().await?;
+    config.connect_timeout = Some(std::time::Duration::from_secs(1));
+    let client = kube::Client::try_from(config)?;
+    Ok(client)
+}
+
+pub(crate) async fn admission_request(
+    operation: Operation,
+    object: Option<PathBuf>,
+    old_object: Option<PathBuf>,
+) -> Result<()> {
+    validate_params(&operation, object.as_ref(), old_object.as_ref())?;
+
+    let output = match operation {
+        Operation::Create => {
+            scaffold_create(
+                RESOURCE_CATALOG_FILE.to_path_buf(),
+                kube_client,
```

Comment:
are we using the correct kube_client here? shouldn't we be using the newly created function?

Reply:
We were using the right function, the naming was not making that evident, though.

I've fixed that with ddd8916b938fe778e42e5c8c5712131ef116d28f